runtime/vm/LayoutFFITypeHelpers.hpp

```diff
@@ -399,6 +400,57 @@ class LayoutFFITypeHelpers
 		return primSigType;
 	}
 
+	/* @brief Compare the struct (rather than the union) size and the specified layout ([FF]/[DD]) size
+	 * in bytes to determine whether they are identical on z/OS or not in the case of ALL_SP/ALL_DP.
+	 *
+	 * @param structSig[in] A pointer to the specified struct signature string
+	 * @param structSize[in] the struct size in bytes
+	 * @param layoutSize[in] the layout size in bytes
+	 * @return true if they are identical on z/OS; false otherwise
+	 */
+	static bool
+	isStructWithFFOrDD(const char *structSig, U_32 structSize, U_32 layoutSize)
+	{
+#if defined(J9ZOS390)
+		/* Check whether the specified signature is a union (starting from '{')
+		 * which must be excluded given a union is treated as a non-complex
+		 * type according to the z/OS ABI.
+		 *
+		 * Note:
+		 * [1] The following float/double related structs ('T' stands for float or double) are complex:
+		 * 1) struct {struct {T a[1]; T b[1];}} which is"[1:F1:F]" or "[1:D1:D]" in the signature string.
+		 * 2) struct {struct {T a[1]; T b;} which is "[1:FF]" or "[1:DD]" in the signature string.
+		 * 3) struct {struct {T a; T b[1];} which is "[F1:F]" or "[D1:D]" in the signature string.
+		 * [2] The following float/double related union/structs are non-complex even though they are
+		 * ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]" in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ((structSize == layoutSize) /* The struct must be 8-byte or 16-byte in size. */
+				&& '{' != structSig[0]) /* A union starts from '{'. */
+				&& !strstr(structSig, "[[FF]]") /* A struct nested by "[[FF]]" or its variant. */
+				&& !strstr(structSig, "[[DD]]") /* A struct nested by "[[DD]]" or its variant. */
+				&& !(('2' == structSig[1]) && (':' == structSig[2])); /* "[2:F]" or "[2:D]" contains "2:" from index 1. */
```

Comment:
i think these tests cannot cover all the positive and negative cases respectively ... these probably are only good for catching a narrow band of them. i am also wondering if here is the right place to catch them all (or to saturate the right testing). 

in my mind, the better place to do this and hopefully catch them all:  somewhere in `parseStruct` ... eventually looking for an outer  struct with two members, 8/16 byte in size,  and isAllSP/isAllDP.

Reply:
> i think these tests cannot cover all the positive and negative cases respectively 

There is no 100% guarantee that the fix covers all corner cases which totally depend upon how they are handled on z/OS against the z/OS ABI. But the fix should already handle what explained by @r30shah as above.

Reply:
> the better place to do this and hopefully catch them all: somewhere in parseStruct ... eventually looking for an outer struct with two members, 8/16 byte in size, and isAllSP/isAllDP.

Technically it might be possible to do that in `parseStruct` plus `setByteCellforStruct` and other methods called by `parseStruct` but essentially there is no difference and even more complicated/tricky as compared to the current fix because we still have to reply on signature string index to determine what the signature string looks like to handle these corner cases. And there is no way to guarantee that the new solution cover all corner cases even though we can achieve that.

Reply:
I don't think `strstr()` is the right tool for this task.

Reply:
> I don't think `strstr()` is the right tool for this task.

What's your proposal for the replacement of `strstr()`?

Reply:
As talked to @keithc-ca offline to avoid confusion, the fix is the simplest way cover all corner cases we encountered for now.